daft/file.py

```diff
@@ -0,0 +1,125 @@
+    with Python's file protocol. It handles both filesystem-based files and in-memory
+    data through its concrete subclasses PathFile and MemoryFile.
+
+    The File object can be used with most Python libraries that accept file-like objects,
+    and implements the standard read/seek/tell interface. Files are read-only in the
+    current implementation.
+
+    Examples:
+        >>> import daft
+        >>> from daft.functions import file
+        >>> df = daft.from_pydict({"paths": ["data.json"]})
+        >>> df = df.select(file(df["paths"]))
+        >>>
+        >>> @daft.func
+        >>> def read_json(file: daft.File) -> str:
+        >>>     import json
+        >>>     data = json.load(file)
+        >>>     return data["text"]
+    """
+
+    _inner: PyDaftFile
+
+    def __init__(self) -> None:
+        raise NotImplementedError("We do not support creating a File via __init__ ")
+
+    @staticmethod
+    def _from_py_daft_file(f: PyDaftFile) -> File:
+        file = File.__new__(File)
+        file._inner = f
+        return file
+
+    @staticmethod
+    def _from_path(path: str) -> File:
+        inner = PyDaftFile._from_path(path)
+        file = PathFile.__new__(PathFile)
+        file._inner = inner
+        return file
+
+    @staticmethod
+    def _from_bytes(bytes: bytes) -> File:
+        inner = PyDaftFile._from_bytes(bytes)
+        file = MemoryFile.__new__(MemoryFile)
+        file._inner = inner
+        return file
+
+    def read(self, size: int = -1) -> bytes:
+        return self._inner.read(size)
+
+    def seek(self, offset: int, whence: int = 0) -> int:
+        return self._inner.seek(offset, whence)
+
+    def tell(self) -> int:
+        return self._inner.tell()
+
+    def close(self) -> None:
+        self._inner.close()
+
+    def open(self, _io_config: IOConfig | None = None) -> File:
+        raise NotImplementedError("File.open() not yet supported")
+
+    def __enter__(self) -> File:
+        inner = self._inner.__enter__()
+        self._inner = inner
+        return self
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        self._inner.__exit__(exc_type, exc_val, exc_tb)
+
+    def __str__(self) -> str:
+        return self._inner.__str__()
+
+    def readable(self) -> bool:
+        return True
+
+    def writable(self) -> bool:
+        return False
+
+    def seekable(self) -> bool:
+        return True
+
+    def isatty(self) -> bool:
+        return False
+
+    def closed(self) -> bool:
+        return self._inner.closed()
+
+
+class PathFile(File):
+    """File object backed by a filesystem path.
+
+    A concrete implementation of File that represents data stored in the filesystem.
+    In addition to the standard file interface, PathFile implements the os.PathLike
+    protocol via __fspath__, allowing it to be used with functions that expect file paths.
+
+    """
+
+    def __fspath__(self) -> str:
+        return self._inner.__fspath__()
```

Comment:
@universalmind303 what's a use case for this one? If I have a column with `PathFile`s and I'm distributing this across nodes, how is a node going to access a file that's not on its local filesystem?

Reply:
Is the usecase you're thinking about here something where the contents are too large to stay in memory, so they need to be on disk?  What about having something like a spill-able file -- it represents a file, and Daft is responsible for moving it around and keeping it on the disk of each node and only bringing in chunks of it at runtime when the user calls `.read()` on it.